DESIGN.md

```diff
@@ -325,18 +325,158 @@ In the addition of the previous rules, an operator for `Flowable`:
 
 ### Creation
 
-Creation of the various types should be exposed through factory methods that provide safe construction.
+Unlike RxJava 1.x, 2.x base classes are to be abstract, stateless and generally no longer wrap an `OnSubscribe` callback - this saves allocation in assembly time without limiting the expressiveness. Operator methods and standard factories still live as final on the base classes.
+
+Instead of the indirection of an `OnSubscribe` and `lift`, operators are to be implemented by extending the base classes. For example, the `map`
```

Comment:
For `Flowable` base *class* type we have the `rs.Publisher` as the base *interface* type. In principle, you could define base interfaces for `Observable`, `Single` and `Completable`, but I don't see why unless you want to encourage other libraries to implement them differently than in RxJava.

For `Flowable`, the `Operator<T, R>` == `Function<Subscriber<R>, Subscriber<T>>`; the `OnSubscribe<T>` == `Consumer<Subscriber<T>>` ~ `rs.Publisher<T>`.

The plan is that standard operator implementations, such as `FlowableMap` may take an `rs.Publisher<T>` at least (and `Flowable<T>` at most), but you won't be able to combine `FlowableMap` and `ObservableMap` because the `actual` output class differs and needs different calls to `onSubscribe`.